### PR TITLE
Support for class definitions with static member functions.

### PR DESCRIPTION
**Pull request by zygoloid**
This includes being able to define a class that was previously forward-declared, and being able to define a member function out-of-line that was previously declared inside a class.

No support for fields or methods yet, and a class definition doesn't yet cause the class to be treated as a complete type.

### CODE REVIEW

File: toolchain/check/handle_name.cpp [hunk 1]
```diff
@@ -9,19 +9,43 @@
 
 namespace Carbon::Check {
 
+static auto GetAsNameScope(Context& context, SemIR::NodeId base_id)
```
Comment (jonmeow) [inline toolchain/check/handle_name.cpp:14]: Please add a brief comment to explain what this function is doing, e.g.:

```suggestion
// Returns the name scope corresponding to base_id, or nullopt if not a scope.
// On invalid scopes, prints a diagnostic and still returns the scope.
static auto GetAsNameScope(Context& context, SemIR::NodeId base_id)
```

File: toolchain/sem_ir/formatter.cpp [hunk 2]
```diff
@@ -572,6 +582,26 @@ class Formatter {
     }
   }
 
+  auto FormatNameScope(NameScopeId id, llvm::StringRef separator,
+                       llvm::StringRef prefix) -> void {
+    // Name scopes aren't kept in any particular order. Sort the entries before
+    // we print them for stability and consistency.
+    std::vector<std::pair<NodeId, StringId>> entries;
```
Comment (jonmeow) [inline toolchain/sem_ir/formatter.cpp]:
```suggestion
    llvm::SmallVector<std::pair<NodeId, StringId>> entries;
```

Just suggesting for consistency, the rule of thumb to use SmallVector and not vector in toolchain code.

File: toolchain/check/testdata/class/basic.carbon [hunk 3]
```diff
@@ -0,0 +1,63 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// AUTOUPDATE
+
+class Class {
```
Comment (jonmeow) [inline toolchain/check/testdata/class/basic.carbon:7]: Should there be a test for something which requires delayed function definition processing? e.g.:

```
class Class {
  fn F() -> i32 {
    return G();
  }
  fn G() -> i32 {
    return F();
  }
}
```

(even if this test is currently a failure)

**Review by jonmeow — APPROVED** [review]
Looks great!

File: toolchain/sem_ir/formatter.cpp (hunk 2 above)
Comment (zygoloid) [reply toolchain/sem_ir/formatter.cpp]: (This is pre-existing, just moved from below.) Done.

File: toolchain/check/testdata/class/basic.carbon (hunk 3 above)
Comment (zygoloid) [reply toolchain/check/testdata/class/basic.carbon:7]: Sure, added.

**Pull request merged**